robot/basestation/static/js/elements/navigation/localization.js

```diff
@@ -0,0 +1,129 @@
+let rover_heading = 0
+let rover_x = 0
+let rover_y = 0
+let zoom_value = 1
+let MAX_ZOOM = 2
+let MIN_ZOOM = 0.25
+let canvas_pos
+let keys = { //javascript keycodes
+    'r': 82,
+    'w': 87,
+    'a': 65,
+    's': 83,
+    'd': 68
+}
+
+let base_lat
+let base_long
+$(document).ready(() => {
+    // the elements might not be initialized yet, must test
+    base_lat = $('#antenna-latitude').text()
+    base_long = $('#antenna-longitude').text()
+})
+
+function setup() {
+    let canvas = createCanvas(450, 300)
+    canvas.parent('#canvasContainer')
+    canvas_pos = $('#defaultCanvas0').offset()
+    angleMode(DEGREES)
+}
+
+function draw() {
+    // all of the manual controls
+    if (keyIsDown(keys['r'])) { //reset position
+        rover_heading = 0;
+        zoom_value = 1;
+        rover_x = rover_y = 0;
+    }
+    if (keyIsDown(keys['a'])) { // spin ccw
+        rover_heading -= 1
+    }
+    if (keyIsDown(keys['d'])) { // spin cw
+        rover_heading += 1
+    }
+    adjustmap_rotation_deg() // keep rover_heading within 0-360
+    if (keyIsDown(keys['w'])) { // move forward
+        rover_x += sin(rover_heading);
+        rover_y -= cos(rover_heading);
+    }
+    if (keyIsDown(keys['s'])) { // move backward
+        rover_x -= sin(rover_heading);
+        rover_y += cos(rover_heading);
+    }
+
+    // canvas static
+    background(225) // draw background
+    translate(width / 2, height / 2) // make the (0,0) point the center of canvas
+    scale(zoom_value) // zoom to given value
+    draw_grid() // draw the grid for reference
+    circle(0, 0, 10) // draw circle base station
+
+    // canvas dynamic
+    draw_rover()
+
+}
+
+// function to draw the lines and values on canvas
+// this grid can be drawn of any size, doesn't have to follow canvas values
+// i.e. we can make it so that 10px = 1m and make the grid go from -1km to 1km (-10000 to 10000 pixels)
+// shouldn't forget to adjust the scaling min/max so the whole grid can fit into the small canvas if needed
+function draw_grid(message) {
+    stroke(200)
+    fill(120)
+    for (let x = -width; x < width; x += 50) {
+        line(x, -height, x, height)
+        text(x, x + 1, 12)
+    }
+    for (let y = -height; y < height; y += 50) {
+        line(-width, y, width, y)
+        text(y, 1, y + 12)
+    }
+}
+
+function draw_rover() {
+    check_rover_position() // get the heading, lat, and long values from ROS topic
```

Comment:
Somewhat the opposite should occur. In ros-helpers.js, a subscriber to /rover_position is created and a callback function is defined. That callback function occurs every time a new message is published to /rover_position, and it should somehow trigger a draw_rover() call and pass in the latitude, longitude and heading from the message object the callback received.